### ISSUE TITLE
np.round optional arguments aren't optional

### ISSUE THREAD

**Issue by synapticarbors** [numba/numba#4439]
I was answering a question on Stackoverflow, where someone was having an issue with using `np.round` in the function they were attempting to jit. It is not obvious at all from the documentation, but if you want to round an array, you must use all 3 arguments. I had to look at the source code to figure this out. So if you try the following, numba fails to compile the code:

```python
@nb.jit(nopython=True)
def func(x):
    return np.round(x)

a = np.random.normal(size=(5,)) + 10.0
func(a)
```

with the error:

```
TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Invalid use of Function(<function round_ at 0x10f5e9510>) with argument(s) of type(s): (array(float64, 1d, C))
 * parameterized
In definition 0:
    All templates rejected with literals.
In definition 1:
    All templates rejected without literals.
This error is usually caused by passing an argument of a type that is unsupported by the named function.
[1] During: resolving callee type: Function(<function round_ at 0x10f5e9510>)
[2] During: typing of call at <ipython-input-19-51b877e3f95c> (3)
```

However the following works:

```python
nb.jit(nopython=True)
def func(x):
    out = np.empty_like(x)
    np.round(x, 0, out)
    return out
```

This came up previously in the following closed issue: https://github.com/numba/numba/issues/2648

It seems that at a minimum this should be clarified in the docs, but better, optional arguments as specified by numpy shouldn't be mandatory.

**Comment by stuartarchibald** [comment]
Thanks for the report. First, thanks for taking the time to answer StackOverflow questions, it's much appreciated. Also, I agree, the fix here should be to just make it work!

**Comment by stuartarchibald** [comment]
@synapticarbors would you be interested in trying to provide a fix for this? Seems like `@numba.extending.overload` ought to be sufficient to plug the gaps.

**Comment by synapticarbors** [comment]
I can take a look. Do you mean to take the code [here](https://github.com/numba/numba/blob/92dbdb90e34402f99d158c035f5a1806378161d3/numba/typing/npydecl.py#L1099-L1128) and replace the entire thing with an `@overload` function, or just the bit that falls under `isinstance(out, types.Array)`? I might need some guidance since I'm not familiar with the `AbstractTemplate`.

**Comment by stuartarchibald** [comment]
That'd be great, thanks. The code you note is using the old style split typing and implementation behaviour, for reference, seems like the implementation part is here: https://github.com/numba/numba/blob/92dbdb90e34402f99d158c035f5a1806378161d3/numba/targets/arraymath.py#L2516-L2527

I think to start, just removing the typing and implementation parts for the `type.Array` type and switching that for an `@overload` implementation would be worth a try. Once that's working I'd hope fixing up the rest of it might be easier/it'll be clearer how to do that. If you hit any problems please shout!

**Comment by synapticarbors** [comment]
@stuartarchibald -- if we hypothetically wanted to replace all of the existing machinery with an `@overload` function and a series of type-specific implementations, is there a way to do that with the llvm intrinsic?

https://github.com/numba/numba/blob/92dbdb90e34402f99d158c035f5a1806378161d3/numba/targets/arraymath.py#L2445-L2454

I looked for other examples, but could not find any.

**Comment by stuartarchibald** [comment]
Yes, there is :), `@intrinsic`, it's part of the public `numba.extending` API, https://github.com/numba/numba/blob/92dbdb90e34402f99d158c035f5a1806378161d3/numba/extending.py#L303-L338 (needs documenting in the Numba docs, but the docstring is good!!).

There's a small example of it in use here: https://github.com/numba/numba/blob/92dbdb90e34402f99d158c035f5a1806378161d3/numba/targets/hashing.py#L124-L130, where float32 is being [`fpext`'d](https://llvm.org/docs/LangRef.html#fpext-to-instruction) to a float64.

and a much bigger one here (tuple hash):
https://github.com/numba/numba/blob/92dbdb90e34402f99d158c035f5a1806378161d3/numba/targets/hashing.py#L280-L307

It's also worth having a look in the unicode impl where there's extensive use of `@intrinsic` https://github.com/numba/numba/blob/master/numba/unicode.py

If you would like some more examples or need more help please shout! Also, the Numba core developers are also available on [gitter.im](https://gitter.im/numba/numba) real time chat if that's more convenient. Thanks!

**Comment by bcbnz** [comment]
@stuartarchibald -- I think this can also be closed as fixed by #8234. Again, the example in the original report now works for me on git main.

**Comment by stuartarchibald** [comment]
Thanks for raising this @bcbnz, this is indeed fixed by #8234, closing as fixed.
